Source/ASRunLoopQueue.mm

```diff
@@ -469,3 +489,230 @@ - (void)unlock
 }
 
 @end
+
+#pragma mark - ASCATransactionQueue
+
+@interface ASCATransactionQueue () {
+  CFRunLoopRef _runLoop;
+  CFRunLoopSourceRef _runLoopSource;
+  CFRunLoopObserverRef _runLoopObserver;
+  NSPointerArray *_internalQueue; // Use NSPointerArray so we can decide __strong or __weak per-instance.
+  ASDN::RecursiveMutex _internalQueueLock;
+  BOOL _disableInterfaceStateCoalesce;
+
+  // In order to not pollute the top-level activities, each queue has 1 root activity.
+  os_activity_t _rootActivity;
+
+#if ASRunLoopQueueLoggingEnabled
+  NSTimer *_runloopQueueLoggingTimer;
+#endif
+}
+
+@end
+
+@implementation ASCATransactionQueue
+
+// CoreAnimation commit order is 2000000, the goal of this is to process shortly beforehand
+// but after most other scheduled work on the runloop has processed.
+static int const kASASCATransactionQueueOrder = 1000000;
```

Comment:
We can always update this later, since the current version of the diff brings obvious benefits and has been productionized, but in an ideal world would we not want to do this work at the end of the layout phase of the commit?

For example, collection views will not update their visible cells until the layout pass, so we'd have to wait for the next go-round for the interface state to be updated to reflect the truth. For a ridiculous case, imagine I have a cell node that is red whenever it's invisible and blue whenever it's visible. The current implementation would show red for 1 frame and then jump to blue.

One weird but viable option is for us to do something like this in the observer:

```objc
while (true) {
  // Layout visible windows (CA is about to do this anyway)
  for (window in app.windows) {
    if (!window.hidden) {
      [window layoutIfNeeded];
    }
  }

  // If we have no work to do, break.
  if ([self isEmpty]) {
    break;
  }

  // Otherwise notify observers and loop again.
  [self processQueue];
}
```

That way any layout invalidations that happen due to interface states, and any interface state changes due to layout, are all settled. We would also use a loop counter to limit the maximum recursive dirtying (just like CA).